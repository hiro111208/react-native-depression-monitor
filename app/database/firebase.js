//import firebase from "firebase/app";
import "firebase/firestore";
import firebase from 'firebase'
require('firebase/auth')

const firebaseConfig = {
  apiKey: "AIzaSyDwOUkdxC7kNNILoFTSx05PTUxRci_kcaw",
  authDomain: "fireship-tutorial-646fc.firebaseapp.com",
  projectId: "fireship-tutorial-646fc",
  storageBucket: "fireship-tutorial-646fc.appspot.com",
  messagingSenderId: "289064122528",
  appId: "1:289064122528:web:ee39abfa2183c8499071c5",
  measurementId: "G-D35RDH9V0E",
};

firebase.initializeApp(firebaseConfig);

export default firebase;
