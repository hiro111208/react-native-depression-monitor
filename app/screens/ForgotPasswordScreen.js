import React, { useState } from "react";
import {
    StyleSheet,
    View,
    Text,
    TouchableOpacity,
    TextInput,
    Alert,
    Image,
} from "react-native";
import firebase from "../database/firebase";

import colors from "../config/colors";
import * as indexStyles from "../config/indexStyles";
import DismissKeyboard from "../config/DismissKeyboard";

/*
 * Screen where users can try to reset their passwords if forgotten
 */
function ForgotPasswordScreen(props) {
    const [email, setEmail] = useState("");
    const [message, setMessage] = useState("");

    // Send reset password email to the email provided if there is an account associated to it
    const forgotPassword = () => {
        if (email === "") {
            Alert.alert("Enter an email!");
        } else {
            firebase
                .auth()
                .sendPasswordResetEmail(email)
                .then(() => {
                    setMessage(
                        "You should have received an email to change your password"
                    );
                })
                .catch((error) => setMessage(error.message));
        }
    };

    // Render forgot password screen interface
    return (
        <DismissKeyboard>
            {/*Render field to input user email*/}
            <View style={styles.container}>
                <Image
                    resizeMode={"contain"}
                    style={indexStyles.loginImage}
                    source={require("../assets/hand-logo.png")}
                />
                <View style={styles.forgotFormContainer}>
                    <TextInput
                        style={styles.inputStyle}
                        placeholder="Email"
                        value={email}
                        onChangeText={(val) => setEmail(val)}
                        testID={"TEST_ID_EMAIL_INPUT"}
                    />

                    {/*Render button that calls forgotPassword method */}
                    <TouchableOpacity
                        activeOpacity={0.5}
                        style={indexStyles.loginButton}
                        onPress={() => forgotPassword()}
                        testID={"TEST_ID_FORGOT_BUTTON"}
                    >
                        <Text style={indexStyles.loginButtonText}>
                            RESET PASSWORD
                        </Text>
                    </TouchableOpacity>

                    {/*Render button that allows user to go back to login screen*/}
                    <Text
                        style={styles.textButton}
                        onPress={() => props.navigation.navigate("LoginScreen")}
                    >
                        Back to Login
                    </Text>

                    {/*Render text that shows error or success messages */}
                    <Text testID={"TEST_ID_MESSAGE"} style={{ color: "red" }}>
                        {message}
                    </Text>
                </View>
            </View>
        </DismissKeyboard>
    );
}
export default ForgotPasswordScreen;

const styles = StyleSheet.create({
    container: {
        flex: 1,
        padding: "8%",
        top: "5%",
    },
    inputStyle: {
        width: "100%",
        marginBottom: 15,
        paddingBottom: 15,
        alignSelf: "center",
        borderColor: "#ccc",
        borderBottomWidth: 1,
    },
    textButton: {
        color: colors.darkBorder,
        marginTop: 25,
        textAlign: "center",
    },
});
