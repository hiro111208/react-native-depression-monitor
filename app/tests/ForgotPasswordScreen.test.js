import firebase from '../../firebase.js';
import React from 'react';
import ForgotPasswordScreen from '../screens/ForgotPasswordScreen';
import renderer from 'react-test-renderer';
import { Alert } from 'react-native';
import {cleanup, render, fireEvent, waitFor} from '@testing-library/react-native';
import "@testing-library/jest-dom/extend-expect";


describe('forgot your password', () => {
    afterEach(cleanup);

test('renders correctly', () => {
    const tree = renderer.create(<ForgotPasswordScreen />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('sendPasswordResetEmail should be unsuccessful with no account related to email', async () => 
{
    let error = '';
    try {
        await firebase.auth().sendPasswordResetEmail('unregistered@test.com');
    } catch (err) {
        error = err.message;
    }
    expect(error).toEqual('There is no user record corresponding to this identifier. The user may have been deleted.');
},10000);

test('sendPasswordResetEmail should be successful with email related to registered account', async () => 
{
    let error= '';
    await firebase.auth().createUserWithEmailAndPassword('registered@test.com', 'password');
    try {
        await firebase.auth().sendPasswordResetEmail('registered@test.com');
    } catch (err) {
        error = err.message;
    }
    await firebase.auth().currentUser.delete().catch(error => {
        credential = firebase.auth.EmailAuthProvider.credential('registered@test.com','password');
        firebase.auth().currentUser.reauthenticateWithCredential(credential)
        firebase.auth().currentUser.delete()
    });
    expect(["","Exceeded daily quota for resetting passwords."]).toContain(error)
},10000);

test('Expect to show email required', async()=>{
    const { getByTestId } = render(<ForgotPasswordScreen />)
    const emailInput= getByTestId('TEST_ID_EMAIL_INPUT');
    const button= getByTestId('TEST_ID_FORGOT_BUTTON');

    jest.spyOn(Alert, 'alert');
    await waitFor(()=> {
        fireEvent.press(button);

        expect(Alert.alert).toHaveBeenCalledWith('Enter an email!');
    })
},10000)

test('Expect to call sendPasswordResetEmail', async()=>{
    const { getByTestId } = render(<ForgotPasswordScreen />);
    const {getByText}= render(<ForgotPasswordScreen />);
    const emailInput= getByTestId('TEST_ID_EMAIL_INPUT');
    const button= getByTestId('TEST_ID_FORGOT_BUTTON');
    const message= getByTestId('TEST_ID_MESSAGE');

    await waitFor(() => button);
    await firebase.auth().createUserWithEmailAndPassword('registered@test.com', 'password');

    jest.spyOn(firebase.auth(), 'sendPasswordResetEmail');
    await fireEvent.changeText(emailInput, 'registered@test.com');
    await fireEvent.press(button);
    
    await waitFor(() => {
        expect(firebase.auth().sendPasswordResetEmail).toHaveBeenCalled();
        expect(["You should have received an email to change your password","Exceeded daily quota for resetting passwords."]).toContain(getByTestId('TEST_ID_MESSAGE').props.children) 
    });
    
    await firebase.auth().currentUser.delete().catch(error => {
        credential = firebase.auth.EmailAuthProvider.credential('registered@test.com','password');
        firebase.auth().currentUser.reauthenticateWithCredential(credential)
        firebase.auth().currentUser.delete()
    });
},10000)
});