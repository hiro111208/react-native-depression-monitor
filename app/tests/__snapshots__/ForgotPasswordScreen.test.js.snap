// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ForgotPasswordScreen.js ForgotPasswordScreen renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "display": "flex",
        "flex": 1,
        "padding": 35,
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Email"
    rejectResponderTermination={true}
    style={
      Object {
        "alignSelf": "center",
        "borderBottomWidth": 1,
        "borderColor": "#ccc",
        "marginBottom": 15,
        "marginTop": 15,
        "paddingBottom": 20,
        "width": "100%",
      }
    }
    testID="TEST_ID_EMAIL_INPUT"
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#ffa351ff",
        "borderRadius": 50,
        "marginTop": 10,
        "opacity": 1,
        "paddingBottom": 15,
        "paddingTop": 15,
        "width": 300,
      }
    }
    testID="TEST_ID_FORGOT_BUTTON"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      SUBMIT
    </Text>
  </View>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "#ffa351ff",
        "marginTop": 25,
        "textAlign": "center",
      }
    }
  >
    Back to Login
  </Text>
  <Text
    style={
      Object {
        "color": "red",
        "flexWrap": "wrap",
        "fontSize": 15,
        "textAlign": "center",
      }
    }
    testID="TEST_ID_MESSAGE"
  >
    
  </Text>
</View>
`;
